op/model.py

```diff
@@ -186,26 +206,34 @@ def __getitem__(self, key):
 
 # We mix in MutableMapping here to get some convenience implementations, but whether it's actually
 # mutable or not is controlled by the flag.
-class RelationUnitData(LazyMapping, MutableMapping):
-    def __init__(self, relation, unit, is_mutable, backend):
+class RelationEntityData(LazyMapping, MutableMapping):
+    def __init__(self, relation, entity, is_mutable, backend):
         self.relation = relation
-        self.unit = unit
+        self.entity = entity
         self._is_mutable = is_mutable
         self._backend = backend
 
+        if isinstance(self.entity, Application):
+            self._is_app = True
+        else:
+            self._is_app = False
+
     def _load(self):
         try:
-            return self._backend.relation_get(self.relation.id, self.unit.name)
+            return self._backend.relation_get(self.relation.id, self.entity.name, self._is_app)
         except RelationNotFound:
             # Dead relations tell no tales (and have no data).
             return {}
 
     def __setitem__(self, key, value):
-        if not self._is_mutable:
-            raise RelationDataError(f'cannot set relation data for {self.unit.name}')
+        if not self._is_mutable(self._backend):
```

Comment:
There's actually a way around that. We can cache the time we did the call, and then prevent calling it again within that timeframe. This is safe as long as the lease time doesn't change, and as long as we check the time _before_ we do the call (so the window is guaranteed correct by the ordering).

Reply:
Ah, and ideally the clock used would be monotonic. How is Python doing in that regard nowadays?

Reply:
[`time.monotonic()`](https://docs.python.org/3/library/time.html#time.monotonic) is in the stdlib (since 3.3, with improvements in 3.5), so it seems like it should be reliable.

Reply:
Before finalizing this, it is worth finding out what happens if leader-set was successfully called before losing leadership. Do the leader-set changes silently roll back? Or does the hook fail? I suspect the hook fails if leadership is lost mid-hook (or it should). Also need to find out if leader-set renews the lease, or if the charm has no way of knowing how old the lease is.

(I don't think any charms worry about it and will just fail, since it involves a netsplit happening when running a hook that takes >30s to run)

Reply:
For reference, here is the code path from the is-leader hook to to the `ClaimLeader` call in the leadership tracker.

https://github.com/juju/juju/blob/juju-2.7-rc6/worker/uniter/runner/jujuc/is-leader.go#L47
https://github.com/juju/juju/blob/juju-2.7-rc6/worker/uniter/runner/context/leader.go#L55-L68 (`ensureLeader` is called here).
https://github.com/juju/juju/blob/juju-2.7-rc6/worker/uniter/runner/context/leader.go#L103-L113
https://github.com/juju/juju/blob/juju-2.7-rc6/worker/leadership/tracker.go#L109-L112

Based on what we discussed so far, my understanding is that we will not use `leader-set` and instead use `relation-set --app` making the framework depend on Juju 2.7 as a base.

> if leader-set renews the lease

`leader-set` also makes sure that the current unit is a leader by calling `ensureLeader`.

https://github.com/juju/juju/blob/juju-2.7-rc6/worker/uniter/runner/context/leader.go#L70-L76
The test for leader-set covers the failure scenario so an error is returned when 

Which is also covered by this test scenario:
https://github.com/juju/juju/blob/juju-2.7-rc6/worker/uniter/runner/context/leader_test.go#L200-L215

> Do the leader-set changes silently roll back?

If `leader-set` returns successfully, it means that all changes were successfully committed to the controller db during its execution and the changes will not be rolled back, i.e. `leader-set` changes get committed immediately but `leader-get` is cached at the unit agent side.

I brought this up as a discussion point during our first sprint and it was decided that we will move to peer application relation data instead to have identical commit semantics for all data bags.

Reply:
@niemeyer @johnsca @jameinel 

Added another commit to this PR to cover temporary caching of is-leader return values as suggested.

Reply:
Moved this to a separate PR:

https://github.com/canonical/operator/pull/59